pkg/tls/tls_test.go

```diff
@@ -0,0 +1,87 @@
+package tls
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"net"
+	"testing"
+	"time"
+)
+
+// x509 expects quite a large number so 0xDEADBEEF will not cut it
+func getSerialNumber() (*big.Int, error) {
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		fmt.Errorf("failed to create cert serial number")
+	}
+	return serialNumber, nil
+}
+func getTemplate() (*x509.Certificate, error) {
+	now := time.Now()
+	SN, err := getSerialNumber()
+	if err != nil {
+		fmt.Errorf("Failed to create cert serial number")
+	}
+
+	template := &x509.Certificate{
+		SerialNumber:          SN,
+		NotBefore:             now,
+		NotAfter:              now.AddDate(5, 0, 0),
+		Subject:               pkix.Name{CommonName: cn},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA:                  true,
+	}
+	return template, nil
+}
+
+var cn = "dummy-test-certificate.local"
+
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	if newPK.Validate() != nil {
+		t.Errorf("Failed to validate private key")
+	}
+	// TODO: verify length of key
+}
+
+func TestNewCA(t *testing.T) {
+	pk, cert, err := NewCA(cn)
+	if err != nil {
+		t.Errorf("failed to create new CA %s", err)
+	}
+	if pk.Validate() != nil {
+		t.Errorf("failed to validate RSA Private key")
+	}
+	opts := x509.VerifyOptions{DNSName: cn}
+	if _, err := cert.Verify(opts); err != nil {
+		t.Errorf("Failed to verify created self signed certificate %s", err)
+	}
+}
+
+func TestNewSelfSignedCert(t *testing.T) {
+	cn := cn
+	org := []string{"Red Hat Internal"}
+	dnsNames := []string{cn}
+	ipAddresses := []net.IP{net.ParseIP("192.168.0.1")}
+	keyUsage := []x509.ExtKeyUsage{x509.ExtKeyUsageAny}
+	selfSign := true
+
+	pk, cert, err := NewCert(cn, org, dnsNames, ipAddresses, keyUsage, nil, nil, selfSign)
+	if err != nil {
+		t.Errorf("error creating new cert %s", err)
+	}
+	if pk.Validate() != nil {
```

Comment:
I think this will fail always as its SelfSign. In self sign case it should be signed by itself. In this case we would need to check if `issued By: Common Name (CN)	*.badssl.com` matches `Issued to: Common Name (CN)	*.badssl.com` and is what we provided.